gmail/gmail_connector.bal

```diff
@@ -0,0 +1,735 @@
+// Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package gmail;
+
+import ballerina/io;
+import ballerina/mime;
+import ballerina/net.http;
+import ballerina/net.uri;
+import oauth2;
+
+@Description {value:"Struct to define the Gmail Client Connector"}
+public struct GmailConnector {
+    oauth2:OAuth2Endpoint oauthEndpoint;
+    string baseUrl;
+}
+
+@Description {value:"List the messages in user's mailbox"}
+@Param {value:"userId: The user's email address. The special value *me* can be used to indicate the authenticated user"}
+@Param {value:"filter: SearchFilter struct with optional query parameters"}
+@Return {value:"MessageListPage struct with array of messages, size estimation and next page token"}
+@Return {value:"GmailError is thrown if any error occurs in sending the request and receiving the response"}
+public function <GmailConnector gmailConnector> listAllMails (string userId, SearchFilter filter) returns (MessageListPage)|GmailError {
+    endpoint oauth2:OAuth2Endpoint oauthEP = gmailConnector.oauthEndpoint;
+    http:Request request = {};
+    http:Response response = {};
+    http:HttpConnectorError connectionError = {};
+    GmailError gmailError = {};
+    string getListMessagesPath = USER_RESOURCE + userId + MESSAGE_RESOURCE;
+    string uriParams = "";
+    //Add optional query parameters
+    uriParams = uriParams + INCLUDE_SPAMTRASH + filter.includeSpamTrash;
```

Comment:
If includeSpamTrash is optional parameter, then don't you need to do the empty check? If it is mandatory one then you don't need the empty check in line no 60.

Reply:
includeSpamTrash parameter is a field in SearchFilter struct which is a boolean. Since in ballerina by default boolean values have false, I am not doing an empty check here. If the user has not set that field, then it will be passed to this as false.

Reply:
In that case, you don't need the empty check in https://github.com/wso2-ballerina/package-gmail/pull/2/files#diff-6202293ec80c29c590212c2178871846R54 as uriParams will always have the vale.

Reply:
Added changes for that